src/rbus/rbus.c

```diff
@@ -5058,4 +5311,498 @@ rbusError_t rbusHandle_GetTraceContextAsString(
     return RBUS_ERROR_SUCCESS;
 }
 
+void rbusRunnableQueue_Init(rbusRunnableQueue_t *q)
+{
+    q->head = NULL;
+    q->tail = NULL;
+    pipe2(q->pipe_fds, O_CLOEXEC);
+    pthread_mutex_init(&q->mutex, NULL);
+    pthread_cond_init(&q->cond, NULL);
+}
+
+void rbusRunnableQueue_PushBack(rbusRunnableQueue_t *q, rbusRunnable_t r)
+{
+    rbusRunnable_t *item = rt_malloc(sizeof(rbusRunnable_t));
+    item->argp = r.argp;
+    item->exec = r.exec;
+    item->cleanup = r.cleanup;
+    item->next = NULL;
+    item->id = r.id;
+
+    pthread_mutex_lock(&q->mutex);
+    if (!q->head)
+    {
+        q->head = item;
+        q->tail = item;
+    }
+    else
+    {
+        q->tail->next = item;
+        q->tail = item;
+    }
+    pthread_mutex_unlock(&q->mutex);
+
+    static const char kQueueIndicator_Runnable = 'r';
+    write(q->pipe_fds[1], &kQueueIndicator_Runnable, 1);
+}
+
+rbusRunnable_t *
+rbusRunnableQueue_PopFront(rbusRunnableQueue_t *q)
+{
+    rbusRunnable_t *r = NULL;
+
+    pthread_mutex_lock(&q->mutex);
+    if (q->head)
+    {
+        r = q->head;
+        q->head = q->head->next;
+    }
+    pthread_mutex_unlock(&q->mutex);
+
+    if (r)
+    {
+        char queue_indicator = 'z';
+        read(q->pipe_fds[0], &queue_indicator, 1);
+    }
+
+    return r;
+}
+
+int
+rbusHandle_GetEventFD(rbusHandle_t rbus)
+{
+    return rbus->eventQueue.pipe_fds[0];
+}
+
+rbusError_t
+rbusHandle_Run(rbusHandle_t rbus)
+{
+  rbusRunnable_t* curr;
```

Comment:
Do you think having a re-entrancy sanity check makes sense here (and in RunOne)?

The Run() APIs can be called both explicit and implicitly, with the latter being rbusAsyncRequest_WaitUntil. If someone sets up code that tries to call WaitUntil within a runnable's callback, it'll lead to some nasty bugs, including deadlocks (around the sync points where rtFuture_Wait is used now). While I can agree that "this is illegal", it will happen and without any safeguards it'll be really painful to debug this.

A simple atomic / semaphore would work. It could either return a special error code, but I'm actually more inclined into making it abort the entire program. There's no easy way to implicitly log this without getting lost somehow.

Reply:
A separate checkin will be made to address on top of this change.